Tools/decrypt_ulog.py

```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+
+from Crypto.PublicKey import RSA
+from Crypto.Cipher import PKCS1_OAEP
+from Crypto.Cipher import ChaCha20
+from Crypto.Hash import SHA256
+import binascii
+import argparse
+#from pathlib import Path
+import sys
+
+if(__name__ == "__main__"):
+
+    parser = argparse.ArgumentParser(description="""CLI tool to decrypt an ulog file\n""")
+    parser.add_argument("ulog_file", help=".ulog file", nargs='?', default=None)
+    parser.add_argument("ulog_key", help=".ulogk, encrypted key", nargs='?', default=None)
+    parser.add_argument("rsa_key", help=".pem format key for decrypting the ulog key", nargs='?', default=None)
+
+    args = parser.parse_args()
+
+    # Only generate a key pair, don't sign
+    if not args.ulog_file or not args.ulog_key or not args.rsa_key:
+        print('Need all arguments, the encrypted ulog file, the key and the key decryption key')
+        sys.exit(1);
+
+    # Read the private RSA key to decrypt the cahcha key
+    with open(args.rsa_key, 'rb') as f:
+        r = RSA.importKey(f.read(), passphrase='')
+
+    # Read the encrypted xchacha key and the nonce
+    with open(args.ulog_key, 'rb') as f:
```

Comment:
Often the nonce is attached directly in front of the cipher (that's what python's ChaCha20 actually assumes if you don't pass the nonce separately), and the key is just the key without any extra information. In this case is it more convenient to deliver the nonce together with the encrypted key. It doesn't rmake any difference really

Normally file magic+content description is not used, since the one decrypting the data knows this anyway (he has the knowledge of algorithm, file format and has access to the private key). So magic + content information is not strictly needed in the file format.

Not having the magic+content in the files adds a layer of obscurity for outsiders who are not supposed to work with the data. I don't care about obscurity, so we can add this info if you insist :)

And of course you can think of an use case where someone, who is authorized to decrypt data, receives logs with different crypto keys & algorithms and needs to select from those when de-crypting.

Just let me know if this is something you really want to have in this PR!

Reply:
> And of course you can think of an use case where someone, who is authorized to decrypt data, receives logs with different crypto keys & algorithms and needs to select from those when de-crypting.

Yes these are the things I'm thinking about. To me a binary file should always be identifiable and versioned. It avoids misinterpretation, user mistakes, and allows future changes.

Reply:
Understood, I will add a header next time I come back to this PR :)

Reply:
@bkueng I drafted a header into key exchange file (header + encrypted key + nonce). Is this the kind of header that you were after? It is on topmost commit.

Reply:
Yes, exactly. Can you also check here for the file version and that the algorithm matches the expected value?

Reply:
Thanks, updated the PR. Changed the version from ascii '0' to just a byte and made it explicit in the header. Also in the python code, checked that the exchange algorithm is the expected one